sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryParallelReadTableSource.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.options.ValueProvider.NestedValueProvider;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An abstract {@link BoundedSource} to read a table from BigQuery.
+ *
+ * <p>This source is to read Bigquery data directly using ParallelRead API.
+ * <ul>
+ * <li>{@link BigQueryParallelReadTableSource} is for reading BigQuery tables</li>
+ * </ul>
+ * ...
+ */
+class BigQueryParallelReadTableSource<T> extends BoundedSource<T> {
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryParallelReadTableSource.class);
+
+  private BigQueryServices bqServices;
+  private Coder<T> coder;
+  private SerializableFunction<SchemaAndRowProto, T> parseFn;
+  private final AtomicReference<Long> tableSizeBytes;
+  private Session session;
+  private ValueProvider<String> jsonTable;
+  private ReadLocation initialReadLocation;
+  private ReadSessionOptions readOptions;
+  private transient List<BoundedSource<T>> cachedSplitResult;
+
+  private BigQueryParallelReadTableSource(ValueProvider<String> jsonTable,
+                                Session session,
+                                ReadLocation readLocation,
+                                SerializableFunction<SchemaAndRowProto, T> parseFn,
+                                Coder<T> coder,
+                                BigQueryServices bqServices,
+                                ReadSessionOptions readOptions) {
+    this.bqServices = checkNotNull(bqServices, "bqServices");
+    this.coder = checkNotNull(coder, "coder");
+    this.parseFn = checkNotNull(parseFn, "parseFn");
+    this.jsonTable = checkNotNull(jsonTable, "jsonTable");
+    this.session = session;
+    this.tableSizeBytes = new AtomicReference<>();
+    this.initialReadLocation = readLocation;
+    this.readOptions = readOptions;
+  }
+
+  /**
+   * This method creates a {@link BigQueryParallelReadTableSource} with no initial session or read
+   * location.
+   */
+  public static <T> BigQueryParallelReadTableSource<T> create(
+      ValueProvider<TableReference> table,
+      SerializableFunction<SchemaAndRowProto, T> parseFn,
+      Coder<T> coder,
+      BigQueryServices bqServices,
+      ReadSessionOptions readOptions) {
+    return new BigQueryParallelReadTableSource<>(
+        NestedValueProvider.of(checkNotNull(table, "table"), new TableRefToJson()),
+        null,
+        // First TableSource shouldn't be used to CreateReader.
+        ReadLocation.getDefaultInstance(),
+        parseFn,
+        coder,
+        checkNotNull(bqServices),
+        readOptions);
+  }
+
+  /*
+   * Gets the total size of the table.
+   */
+  @Override
+  public synchronized long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    if (tableSizeBytes.get() == null) {
```

Comment:
Is it guaranteed to be `.isAccessible()` when this method is called? I think it isn't - it can be called at construction time. You can avoid such errors by having your unit tests supply everything that can be supplied as a ValueProvider, using TestPipeline.newProvider() instead of StaticValueProvider. It simulates a parameter with a given value available only at runtime. But you'll need to test the "available at construction time" case too, separately.

Reply:
I'm not sure I understand your comment here -- are you referring to the jsonTable provider below? IIRC this is exactly what the existing BigQueryTableSource does.